flowgate-api/src/main/java/com/vmware/flowgate/service/AssetService.java

```diff
@@ -956,4 +988,397 @@ private void filterValueUnitsByMetricNames(List<ValueUnit> valueunits, List<Stri
+         //3. Prepare translate context
+         List<Map<String, String>> contextMap =
+               getTranslateContextMap(assetAndValueUnitsMap, asset);
+         //4. Translate
+      default:
+         assetAndValueUnitsMap = getRawMetrics(asset, starttime, duration);
+         //return translateToMetricDataForOtherAsset(otherValueUnits);
+      }
+      return null;
+   }
+
+   /**
+    *
+    * @param assetAndValueUnitsMap
+    * @param asset
+    * @return
+    * Data structure of the map:
+    * {
+    *    ToName:"",
+    *    Formula:"",
+    *    IDAndValueUnitMap:[
+    *       [{id,valueUnit},{id,valueUnit}]
+    *    ]
+    * }
+    * One Map<String, String> include the ToName and Formula, and [<id,ValueUnit>]:List<Map<id,valueUnit>>
+    * One MetricName may correspond to one or multiple values
+    * When the API of get-latest-data use the method, One MetricName correspond one value
+    * When the API of get-duration-data use the method, One MetricName correspond multiple values
+    */
+   private List<Map<String, String>> getTranslateContextMap(Map<String, List<ValueUnit>> assetAndValueUnitsMap,
+         Asset asset){
+      List<Map<String, String>> contextMapList = new ArrayList<>();
+      Map<String, Map<String, String>> toNameAndFormulasMap = getToNameAndFromulasMapForServer(asset);
+      if(toNameAndFormulasMap.isEmpty()) {
+         return contextMapList;
+      }
+      //For Server Voltage
+      Map<String, List<ValueUnit>> idAndValuesForServerVoltage = new HashMap<>();
+      List<ValueUnit> inletVoltages = new ArrayList<>();
+      List<ValueUnit> outletVoltages = new ArrayList<>();
+
+      CompareValueUnitByTime comparator = new CompareValueUnitByTime();
+      for(Map.Entry<String, Map<String, String>> toNameAndFormulaEntry : toNameAndFormulasMap.entrySet()) {
+         String toName = toNameAndFormulaEntry.getKey();
+         Map<String, String> contextMap = new HashMap<String, String>();
+         Map<String, String> metricNameAndFormulaMap = toNameAndFormulaEntry.getValue();
+         for(Map.Entry<String, String> metricNameAndFormula : metricNameAndFormulaMap.entrySet()) {
+            String metricName = metricNameAndFormula.getKey();
+            String idFormula = metricNameAndFormula.getValue();
+            String ids[] = idFormula.split("\\+|-|\\*|/|\\(|\\)");
+            Map<String, List<ValueUnit>> idAndValues = new HashMap<>();
+            for(String id : ids) {
+               List<ValueUnit> valueUnits = new ArrayList<>();
+               for(ValueUnit valueUnit : assetAndValueUnitsMap.get(id)) {
+                  if(!metricName.equals(valueUnit.getKey())) {
+                     continue;
+                  }
+                  String pdu_outlet_extraidentifier = valueUnit.getExtraidentifier();
+                  String outlet = null;
+                  switch (metricName) {
+                  case MetricName.PDU_CURRENT:
+                  case MetricName.PDU_APPARENT_POWER:
```

Comment:
may not the right place, but we should try to keep valueunit name and metric name as the same value. like pdu total current and current of the the outlet and inlet, they should be both called current, use extra-identifier can identify them .